### PR TITLE
[IA-1813] Inactive zombie monitoring

### PR DESCRIPTION
**Pull request by akarukappadath**
\<your comments for this PR go here\>

---
Have you read [CONTRIBUTING.md](https://github.com/DataBiosphere/leonardo/blob/develop/CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've documented my API changes in Swagger

In all cases:

- [ ] Get a thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; Delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ZombieMonitorQueries.scala
Comment (akarukappadath) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ZombieMonitorQueries.scala]: In order for the runtime to be deleted through a CloudService interpreter, we need to pass in the full runtime, which is why `ZombieCandidate` now has a full Runtime in it. The other option was to have two types of `ZombieCandidate`s - `ActiveZombieCandidate` and `InactiveZombieCandidate` but I thought it made the code in the monitor a lot less easily readable

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ZombieRuntimeMonitor.scala
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ZombieRuntimeMonitor.scala]: would be good to add some metrics..like number of activeZombies (I think we have this already) and number of inactiveZombies

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ZombieRuntimeMonitor.scala
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ZombieRuntimeMonitor.scala]: we probably don't even need to check status in this case...I think as long as the runtime exists in Google, we should delete it whatever state that runtime is in 🤔

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ZombieMonitorQueries.scala
Comment (Qi77Qi) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ZombieMonitorQueries.scala]: Delete really shouldn't need the whole `Runtime`...if you look at both gce and dataproc interpreters, they only really need googleProject, runtimeName, and zoneName (if gce)...Let's refactor `deleteRuntime` signature a bit? :)

**Review by Qi77Qi — COMMENTED** [review]
Looks great 👍 ...Left some minor comments

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ZombieRuntimeMonitor.scala
Comment (akarukappadath) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ZombieRuntimeMonitor.scala]: Yup, they're both in the `handle` function for each

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ZombieMonitorQueries.scala
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ZombieMonitorQueries.scala]: Existing records won't have this label key at all.. does this code take that into account?

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ZombieMonitorQueries.scala
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ZombieMonitorQueries.scala]: Maybe do `join` on runtime configs because we expect every record to have one

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ZombieRuntimeMonitor.scala
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ZombieRuntimeMonitor.scala]: should maybe be `parTraverseFilter` since we're using a `semaphore`?

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ZombieRuntimeMonitor.scala
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ZombieRuntimeMonitor.scala]: Can we make this method return `F[Unit]` and remove this line?

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: Please make the same change in `RuntimeServiceInterp.scala`

**Comment by rtitle** [comment]
Few comments above.. overall the approach looks good, will re-review before merge.

One question: are we expecting this to loop through _all_ currently deleted clusters the first time this runs? There are a lot of them in prod:
```
MySQL [leonardo]> select count(*) from CLUSTER where status = 'Deleted';
+----------+
| count(*) |
+----------+
|    14690 |
+----------+
```
I'm thinking it should be OK since we control concurrency with a Semaphore but we might just want to double check how long it'll take, and make sure we don't blow up any Google quota or anything.

Another approach is just to manually set `deletionConfirmed==true` on all existing deleted clusters via a liquibase migration. Then the new inactive zombie monitoring will only take place for runtimes going forward. We can process all legacy runtimes offline with a script.

**Comment by Qi77Qi** [comment]
@rtitle I was thinking semaphore will protect us, but I think what you're suggesting with migration is probably even better.

**Comment by akarukappadath** [comment]
@rtitle I'm not sure we'd be able to do a liquibase migration for existing runtimes since it's a label (db row) rather than a db column. If we're worried about a google quota, we could write a script to take care of the existing runtimes

**Comment by rtitle** [comment]
@akarukappadath we can run arbitrary SQL in liquibase migrations, example: https://github.com/DataBiosphere/leonardo/blob/develop/http/src/main/resources/org/broadinstitute/dsde/workbench/leonardo/liquibase/changesets/20200317_fix_custom_image_type.xml#L5

Something like
```
INSERT INTO LABEL (resourceId, resourceType, key, value) SELECT id, 'runtime', 'deletionConfirmed', 'true` FROM CLUSTER c where c.status = 'Deleted'
```

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ZombieRuntimeMonitor.scala
Comment (akarukappadath) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ZombieRuntimeMonitor.scala]: I'm worried this might lead to a weird state - what if a runtime is `Deleting` on Google? I think trying to delete a `Deleting` runtime would cause an error right?

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ZombieRuntimeMonitor.scala
Comment (akarukappadath) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ZombieRuntimeMonitor.scala]: Is there a more scala-esque way to do this? `ifA` didn't work here since it looks like it evaluates both parameters and returns the relevant one, which meant that the database was being updated regardless of the boolean value

Comment (Qi77Qi) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ZombieRuntimeMonitor.scala]: I don't think so...you're probably have to use `flatMap` and then if else

File: http/src/main/resources/org/broadinstitute/dsde/workbench/leonardo/liquibase/changesets/20200506_label_pk.xml
Comment (rtitle) [inline http/src/main/resources/org/broadinstitute/dsde/workbench/leonardo/liquibase/changesets/20200506_label_pk.xml]: I'm fine with this change but I'm curious if there is a specific reason for using primary key over unique constraint? Kinda wish I did it in my last PR

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala]: Can you run `scalafmt` and `test:scalafmt`?

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/LabelComponent.scala
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/LabelComponent.scala]: Maybe specify the constraintName in liquibase so the name here matches. Seems you can do:
```
<addPrimaryKey tableName="LABEL" columnNames="resourceId, resourceType, key" contstraintName="PK_LABEL" />
```

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/LabelComponent.scala
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/LabelComponent.scala]: Remove blank line (`scalafmt` might do this for you)

File: http/src/main/resources/org/broadinstitute/dsde/workbench/leonardo/liquibase/changesets/20200506_label_pk.xml
Comment (akarukappadath) [reply http/src/main/resources/org/broadinstitute/dsde/workbench/leonardo/liquibase/changesets/20200506_label_pk.xml]: If we want to use `insertOrUpdate` on the label table (in order to change existing labels) there needs to be a primary key for the table. I slacked about it earlier this week, sorry if I wasn't clear!

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ZombieRuntimeMonitor.scala
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ZombieRuntimeMonitor.scala]: How did we decide on 1 minute? Seems a little aggressive -- I could imaging a 1 minute lag between when front Leo receives the create request and back Leo actually creates the runtime. Maybe 10 minutes would be safer.

File: http/src/main/resources/org/broadinstitute/dsde/workbench/leonardo/liquibase/changesets/20200506_label_pk.xml
Comment (rtitle) [reply http/src/main/resources/org/broadinstitute/dsde/workbench/leonardo/liquibase/changesets/20200506_label_pk.xml]: Yeah I think I realized that, makes sense 👍

**Comment by rtitle** [comment]
Sorry a few more comments, but aside from those this looks great -- all the logic in `ZombieRuntimeMonitor` looks correct to me. Glad we're adding metrics too.

Will thumb once the above is addressed.

**Review by rtitle — APPROVED** [review]
Looks good but as discussed let's hold off merging until after the release tomorrow. Thank you!

**Pull request merged**
